## Soal 1
Ikuti langkah-langkah pada modul praktikum bagian D1 untuk menyelesaikan tugas regresi. Gunakan base estimator Linear Regression dan Decision Tree Regressor dari skLearn dengan import library berikut, masing-masing menggunakan 12 estimator (model dasar) dalam ensemble.


## Linear Regression

In [103]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split #membagi dataset menjadi data latih dan data uji.
from sklearn.ensemble import BaggingRegressor # untuk bagging
from sklearn.linear_model import LinearRegression #memprediksi nilai kontinu berdasarkan hubungan linier antara variabel input dan output.
from sklearn.metrics import mean_squared_error #menghitung Mean Squared Error (MSE), yaitu ukuran untuk menilai seberapa baik model melakukan prediksi (semakin rendah MSE, semakin baik).

In [104]:
#Import dataset
X = np.array([1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5.0, 5.5, 6.0, 6.5, 7.0, 7.5, 8.0]) #Fitur 1 (Fitur, Independen)

# Reshape X agar menjadi array 2D
# Mengubah data dari 1D ( matrix baris) menjadi 2D (matrix kolom)
X = X.reshape(-1, 1)
Y = np.array([2.5, 3.7, 4.1, 5.0, 5.8, 6.5, 7.3, 8.0, 8.9, 9.4, 10.2, 11.1, 11.7, 12.5, 13.4]) # Fitur 2 (Target, Dependen)
print(X)

[[1. ]
 [1.5]
 [2. ]
 [2.5]
 [3. ]
 [3.5]
 [4. ]
 [4.5]
 [5. ]
 [5.5]
 [6. ]
 [6.5]
 [7. ]
 [7.5]
 [8. ]]


In [105]:
# Split data untuk regresi
# Menggunakan fungsi train_test_split dari scikit-learn secara acak
# 25 data uji dan 75 data latih
#random_state = seed
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.25, random_state=42)

In [106]:
print (X_train)
print (Y_train)

[[3.5]
 [5. ]
 [2. ]
 [1.5]
 [8. ]
 [3. ]
 [4.5]
 [6. ]
 [7. ]
 [2.5]
 [4. ]]
[ 6.5  8.9  4.1  3.7 13.4  5.8  8.  10.2 11.7  5.   7.3]


In [107]:
# Bagging Regressor dengan Linear Regression sebagai base estimator
# nilai estimator = banyaknya himpunan (m)
be = LinearRegression()
model = BaggingRegressor (be, n_estimators=12, random_state=42)
model.fit(X_train, Y_train)

BaggingRegressor(estimator=LinearRegression(), n_estimators=12, random_state=42)

In [108]:
# Prediksi dan evaluasi
Y_pred = model.predict(X_test)
mse = mean_squared_error(Y_test, Y_pred)
rmse = np.sqrt(mse)
mae = np.mean(np.abs(Y_pred - Y_test))
r2_score = 1 - (np.sum((Y_pred - Y_test)**2) / np.sum((Y_test - np.mean(Y_test))**2))
print("MSE for Bagging Regressor (Linear Regression):", mse)
print("RMSE for Bagging Regressor (Linear Regression):", rmse)
print("MAE for Bagging Regressor (Linear Regression):", mae)
print("R2 Score for Bagging Regressor (Linear Regression):", r2_score)

MSE for Bagging Regressor (Linear Regression): 0.02782002391886787
RMSE for Bagging Regressor (Linear Regression): 0.1667933569386619
MAE for Bagging Regressor (Linear Regression): 0.12825031193058667
R2 Score for Bagging Regressor (Linear Regression): 0.9981141364118888


In [109]:
#Menampilkan data pada tiap model
print("Data tiap model bootsrap:")
for i, samples in enumerate(model.estimators_samples_, start=1):
  print(f"Data model {i} = {X_train [samples].flatten()}")

Data tiap model bootsrap:
Data model 1 = [2.5 4.  4.  2.  2.5 6.  6.  7.  1.5 2.  7. ]
Data model 2 = [6.  1.5 6.  1.5 2.5 3.  4.  1.5 2.5 4.  3.5]
Data model 3 = [4.5 3.5 6.  6.  2.5 4.  7.  3.  7.  2.  3. ]
Data model 4 = [3.5 3.  6.  4.  8.  5.  1.5 4.5 2.  3.  1.5]
Data model 5 = [4.5 2.5 4.5 1.5 1.5 7.  1.5 6.  7.  4.5 4. ]
Data model 6 = [3.  2.  5.  2.  6.  2.5 1.5 5.  1.5 4.5 5. ]
Data model 7 = [1.5 2.  4.  5.  3.  2.  1.5 5.  3.  3.  4. ]
Data model 8 = [4.5 2.  7.  2.  3.  2.5 3.5 8.  4.  1.5 3. ]
Data model 9 = [7.  4.  8.  8.  1.5 2.  1.5 4.5 4.  7.  1.5]
Data model 10 = [7.  5.  7.  2.5 3.5 6.  4.5 4.  4.  7.  5. ]
Data model 11 = [1.5 2.  2.5 7.  8.  4.5 4.5 1.5 8.  2.  6. ]
Data model 12 = [8.  6.  7.  4.5 4.  6.  5.  2.5 6.  2.  7. ]


In [110]:
#Menampilkan hasil evaluasi tiap model
# Loop untuk menampilkan akurasi setiap estimator/model
print("Akurasi tiap model:")
for i, estimator in enumerate (model.estimators_, start=1):
  # Prediksi menggunakan estimator ke-i
  Y_pred = estimator.predict(X_test)
  mse = mean_squared_error(Y_test, Y_pred)
  print(f"model {i} = {mse:.4f}")

Akurasi tiap model:
model 1 = 0.0234
model 2 = 0.0590
model 3 = 0.0223
model 4 = 0.0317
model 5 = 0.0572
model 6 = 0.0285
model 7 = 0.0345
model 8 = 0.0196
model 9 = 0.0379
model 10 = 0.0362
model 11 = 0.0222
model 12 = 0.0130


## DecisionTreeRegressor

In [111]:
#Import dataset
X = np.array([1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5.0, 5.5, 6.0, 6.5, 7.0, 7.5, 8.0]) #Fitur 1 (Fitur, Independen)

# Reshape X agar menjadi array 2D
# Mengubah data dari 1D ( matrix baris) menjadi 2D (matrix kolom)
X = X.reshape(-1, 1)
Y = np.array([2.5, 3.7, 4.1, 5.0, 5.8, 6.5, 7.3, 8.0, 8.9, 9.4, 10.2, 11.1, 11.7, 12.5, 13.4]) # Fitur 2 (Target, Dependen)
print(X)

[[1. ]
 [1.5]
 [2. ]
 [2.5]
 [3. ]
 [3.5]
 [4. ]
 [4.5]
 [5. ]
 [5.5]
 [6. ]
 [6.5]
 [7. ]
 [7.5]
 [8. ]]


In [112]:
# Split data untuk regresi
# Menggunakan fungsi train_test_split dari scikit-learn secara acak
# 25 data uji dan 75 data latih
#random_state = seed
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.25, random_state=42)
print (X_train)
print (Y_train)

[[3.5]
 [5. ]
 [2. ]
 [1.5]
 [8. ]
 [3. ]
 [4.5]
 [6. ]
 [7. ]
 [2.5]
 [4. ]]
[ 6.5  8.9  4.1  3.7 13.4  5.8  8.  10.2 11.7  5.   7.3]


In [113]:
# Bagging Regressor dengan Linear Regression sebagai base estimator
# nilai estimator = banyaknya himpunan (m)

from sklearn.tree import DecisionTreeRegressor

be = DecisionTreeRegressor()
model = BaggingRegressor (be, n_estimators=12, random_state=42)
model.fit(X_train, Y_train)

BaggingRegressor(estimator=DecisionTreeRegressor(), n_estimators=12,
                 random_state=42)

In [114]:
# Prediksi dan evaluasi
Y_pred = model.predict(X_test)
mse = mean_squared_error(Y_test, Y_pred)
rmse = np.sqrt(mse)
mae = np.mean(np.abs(Y_pred - Y_test))
r2_score = 1 - (np.sum((Y_pred - Y_test)**2) / np.sum((Y_test - np.mean(Y_test))**2))
print("MSE for Bagging Regressor (Decision Tree Regressor):", mse)
print("RMSE for Bagging Regressor (Decision Tree Regressor):", rmse)
print("MAE for Bagging Regressor (Decision Tree Regressor):", mae)
print("R2 Score for Bagging Regressor (Decision Tree Regressor):", r2_score)

MSE for Bagging Regressor (Decision Tree Regressor): 0.9452083333333319
RMSE for Bagging Regressor (Decision Tree Regressor): 0.9722182539601547
MAE for Bagging Regressor (Decision Tree Regressor): 0.8458333333333323
R2 Score for Bagging Regressor (Decision Tree Regressor): 0.9359262240675621


In [115]:
#Menampilkan data pada tiap model
print("Data tiap model bootsrap:")
for i, samples in enumerate(model.estimators_samples_, start=1):
  print(f"Data model {i} = {X_train [samples].flatten()}")

Data tiap model bootsrap:
Data model 1 = [2.5 4.  4.  2.  2.5 6.  6.  7.  1.5 2.  7. ]
Data model 2 = [6.  1.5 6.  1.5 2.5 3.  4.  1.5 2.5 4.  3.5]
Data model 3 = [4.5 3.5 6.  6.  2.5 4.  7.  3.  7.  2.  3. ]
Data model 4 = [3.5 3.  6.  4.  8.  5.  1.5 4.5 2.  3.  1.5]
Data model 5 = [4.5 2.5 4.5 1.5 1.5 7.  1.5 6.  7.  4.5 4. ]
Data model 6 = [3.  2.  5.  2.  6.  2.5 1.5 5.  1.5 4.5 5. ]
Data model 7 = [1.5 2.  4.  5.  3.  2.  1.5 5.  3.  3.  4. ]
Data model 8 = [4.5 2.  7.  2.  3.  2.5 3.5 8.  4.  1.5 3. ]
Data model 9 = [7.  4.  8.  8.  1.5 2.  1.5 4.5 4.  7.  1.5]
Data model 10 = [7.  5.  7.  2.5 3.5 6.  4.5 4.  4.  7.  5. ]
Data model 11 = [1.5 2.  2.5 7.  8.  4.5 4.5 1.5 8.  2.  6. ]
Data model 12 = [8.  6.  7.  4.5 4.  6.  5.  2.5 6.  2.  7. ]


In [116]:
#Menampilkan hasil evaluasi tiap model
# Loop untuk menampilkan akurasi setiap estimator/model
print("Akurasi tiap model:")
for i, estimator in enumerate (model.estimators_, start=1):
  # Prediksi menggunakan estimator ke-i
  Y_pred = estimator.predict(X_test)
  mse = mean_squared_error(Y_test, Y_pred)
  print(f"model {i} = {mse:.4f}")

Akurasi tiap model:
model 1 = 0.8825
model 2 = 2.0450
model 3 = 1.1625
model 4 = 0.8275
model 5 = 0.8825
model 6 = 1.9475
model 7 = 4.8725
model 8 = 1.1000
model 9 = 1.1000
model 10 = 1.9875
model 11 = 0.8825
model 12 = 1.0650


MSE for Bagging Regressor (Linear Regression): 0.02782002391886787      
MSE for Bagging Regressor (Decision Tree Regressor): 0.9452083333333319

## Soal 2
Gunakan dataset dan code pada bagian D3 dengan menambahkan satu fitur baru yaitu 'Cereal_yield_kg_per_hectare_Value' untuk data pelatihan, lakukan klasifikasi dengan skenario 1 menggunakan model Naive Bayes saja tanpa teknik bagging dan skenario 2 menggunakan teknik Bagging dengan base estimator Naive Bayes buat 100 estimator (model dasar) dalam ensemble. Bandingkan hasil akurasi yang diperoleh dari kedua model tersebut. Apakah teknik bagging mampu meningkatkan akurasi model Naive Bayes?


In [117]:
#Import library
import pandas as pd
from sklearn.ensemble import BaggingClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [118]:
#Import dataset
data = pd.read_csv('/content/df_final_features.csv')
data.head(5)

Name Sex   Age  Height  Weight         Team  Year  \
0                 A Dijiang   M  24.0   180.0    80.0        China  1992   
1                  A Lamusi   M  23.0   170.0    60.0        China  2012   
2  Christine Jacoba Aaftink   F  21.0   185.0    82.0  Netherlands  1988   
3  Christine Jacoba Aaftink   F  21.0   185.0    82.0  Netherlands  1988   
4  Christine Jacoba Aaftink   F  25.0   185.0    82.0  Netherlands  1992   

   Season    Host_City    Host_Country  ... GDP_Per_Capita_Constant_LCU_Value  \
0  Summer    Barcelona           Spain  ...                       6875.676999   
1  Summer       London  United Kingdom  ...                      41274.127360   
2  Winter      Calgary          Canada  ...                      24946.565910   
3  Winter      Calgary          Canada  ...                      24946.565910   
4  Winter  Albertville          France  ...                      27485.503400   

  Cereal_yield_kg_per_hectare_Value  Military_expenditure_current_LCU_Value  \
0                            4362.3                            6.849287e+10   
1                            5825.2                            9.935000e+11   
2                            6194.0                            6.035300e+09   
3                            6194.0                            6.035300e+09   
4                            7459.2                            6.307500e+09   

   Tax_revenue_current_LCU_Value  Expense_current_LCU_Value  \
0                   1.605180e+12               7.161170e+12   
1                   5.524090e+12               7.161170e+12   
2                   5.311044e+10               1.125100e+11   
3                   5.311044e+10               1.125100e+11   
4                   6.846182e+10               1.338420e+11   

   Central_government_debt_total_current_LCU_Value  Representing_Host  \
0                                     4.396980e+12                  0   
1                                     3.974130e+13                  0   
2                                     1.441520e+11                  0   
3                                     1.441520e+11                  0   
4                                     1.624580e+11                  0   

   Avg_Temp  Medal Medal_Binary  
0      6.95      0            0  
1      6.95      0            0  
2      9.25      0            0  
3      9.25      0            0  
4      9.25      0            0  

[5 rows x 22 columns]

In [119]:
# Persiapan data
from sklearn.impute import SimpleImputer # Import SimpleImputer

#defined X and Y
X = data[['Age', 'Height', 'Weight', 'Cereal_yield_kg_per_hectare_Value']] #data
Y = data['Sex'] # Target

#Split data
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.25, random_state=42)
#Create an imputer to fill NaN values with the mean of each column
imputer = SimpleImputer (strategy='mean') # or strategy='median'

# Fit the imputer on the training data and transform both training and testing data
X_train = imputer.fit_transform(X_train)
X_test = imputer.transform(X_test)

In [120]:
print (X_train)
print (y_train)

#tampilkan jumlah data latih x (X_train)
print(X_train.shape)

[[  18.         189.          79.        3520.8      ]
 [  21.         161.          63.        5809.5      ]
 [  23.         166.          70.        2243.       ]
 ...
 [  26.         186.          70.        4652.4      ]
 [  36.         181.          82.        3570.9      ]
 [  22.         177.6685237   73.0266582 2550.       ]]
101804    M
71686     F
53571     M
197794    F
129378    F
         ..
119879    F
103694    M
131932    M
146867    M
121958    M
Name: Sex, Length: 151962, dtype: object
(151962, 4)


In [121]:
# Melatih dan evaluasi model tanpa bagging

# model Naive Bayes tanpa bagging
model = GaussianNB()

# Melatih model
model.fit(X_train, y_train)

# Membuat prediksi pada data uji
y_pred = model.predict(X_test)

# Menghitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f"Akurasi Naive Bayes tanpa bagging: {accuracy:.4f}")

Akurasi Naive Bayes tanpa bagging: 0.7892


In [122]:
#Melatih dan evaluasi model dengan bagging Naive Bayes

be_c = GaussianNB()
#Membuat model Bagging dengan Naive Bayes sebagai estimator dasar
bagging_clf = BaggingClassifier(be_c, n_estimators=100, random_state=42)

#Melatih model
bagging_clf.fit(X_train, y_train)

#Membuat prediksi pada data uji
y_pred = bagging_clf.predict(X_test)

#Menghitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f"Akurasi Bagging dengan Naive Bayes: {accuracy:.4f}")

Akurasi Bagging dengan Naive Bayes: 0.7892


In [129]:
#Melatih dan evaluasi model dengan bagging Naive Bayes

be_c = GaussianNB()
#Membuat model Bagging dengan Naive Bayes sebagai estimator dasar
bagging_clf = BaggingClassifier(be_c, n_estimators=200, random_state=42)

#Melatih model
bagging_clf.fit(X_train, y_train)

#Membuat prediksi pada data uji
y_pred = bagging_clf.predict(X_test)

#Menghitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f"Akurasi Bagging dengan Naive Bayes: {accuracy:.4f}")

Akurasi Bagging dengan Naive Bayes: 0.6358


## Soal 3
Gunakan dataset dan code pada bagian D3 dengan menambahkan dua fitur baru yaitu 'Cereal_yield_kg_per_hectare_Value' dan ‘GDP_Per_Capita_Constant_LCU_Value’ , utuk data pelatihan, lakukan klasifikasi dengan skenario 1 menggunakan model Decission Tree saja tanpa teknik bagging. Skenario 2 menggunakan teknik Bagging dengan base estimator Decission Tree buat 100 estimator (model dasar) dalam ensemble. Skenario 3 tingkatkan estimator (model dasar) menjadi 200 atau lebih. Bandingkan hasil akurasi yang diperoleh dari semua model tersebut. Apakah penambahan fitur dan estimator pada teknik bagging mampu meningkatkan akurasi model Decission Tree?


In [123]:
# Persiapan data
from sklearn.impute import SimpleImputer # Import SimpleImputer

#defined X and Y
X = data[['Age', 'Height', 'Weight', 'Cereal_yield_kg_per_hectare_Value', 'GDP_Per_Capita_Constant_LCU_Value']] #data
Y = data['Sex'] # Target

#Split data
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.25, random_state=42)
#Create an imputer to fill NaN values with the mean of each column
imputer = SimpleImputer (strategy='mean') # or strategy='median'

# Fit the imputer on the training data and transform both training and testing data
X_train = imputer.fit_transform(X_train)
X_test = imputer.transform(X_test)

In [124]:
print (X_train)
print (y_train)

#tampilkan jumlah data latih x (X_train)
print(X_train.shape)

[[1.80000000e+01 1.89000000e+02 7.90000000e+01 3.52080000e+03
  2.24112243e+05]
 [2.10000000e+01 1.61000000e+02 6.30000000e+01 5.80950000e+03
  3.62329576e+06]
 [2.30000000e+01 1.66000000e+02 7.00000000e+01 2.24300000e+03
  9.94515166e+03]
 ...
 [2.60000000e+01 1.86000000e+02 7.00000000e+01 4.65240000e+03
  2.30597223e+04]
 [3.60000000e+01 1.81000000e+02 8.20000000e+01 3.57090000e+03
  1.36087874e+04]
 [2.20000000e+01 1.77668524e+02 7.30266582e+01 2.55000000e+03
  9.85962619e+06]]
101804    M
71686     F
53571     M
197794    F
129378    F
         ..
119879    F
103694    M
131932    M
146867    M
121958    M
Name: Sex, Length: 151962, dtype: object
(151962, 5)


In [125]:
# Melatih dan evaluasi model tanpa bagging

# model Decision Tree tanpa bagging
model = DecisionTreeClassifier()

# Melatih model
model.fit(X_train, y_train)

# Membuat prediksi pada data uji
y_pred = model.predict(X_test)

# Menghitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f"Akurasi DecisionTree tanpa bagging: {accuracy:.4f}")

Akurasi DecisionTree tanpa bagging: 0.8633


In [126]:
#Melatih dan evaluasi model dengan bagging

be_c = DecisionTreeClassifier()
#Membuat model Bagging dengan Naive Bayes sebagai estimator dasar
bagging_clf = BaggingClassifier(be_c, n_estimators=100, random_state=42)

#Melatih model
bagging_clf.fit(X_train, y_train)

#Membuat prediksi pada data uji
y_pred = bagging_clf.predict(X_test)

#Menghitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f"Akurasi Bagging dengan DecisionTree: {accuracy:.4f}")

Akurasi Bagging dengan DecisionTree: 0.8943


In [127]:
#Melatih dan evaluasi model dengan bagging

be_c = DecisionTreeClassifier()
#Membuat model Bagging dengan Naive Bayes sebagai estimator dasar
bagging_clf = BaggingClassifier(be_c, n_estimators=200, random_state=42)

#Melatih model
bagging_clf.fit(X_train, y_train)

#Membuat prediksi pada data uji
y_pred = bagging_clf.predict(X_test)

#Menghitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f"Akurasi Bagging dengan DecisionTree: {accuracy:.4f}")

Akurasi Bagging dengan DecisionTree: 0.8952


In [128]:
#Melatih dan evaluasi model dengan bagging

be_c = DecisionTreeClassifier()
#Membuat model Bagging dengan Naive Bayes sebagai estimator dasar
bagging_clf = BaggingClassifier(be_c, n_estimators=500, random_state=42)

#Melatih model
bagging_clf.fit(X_train, y_train)

#Membuat prediksi pada data uji
y_pred = bagging_clf.predict(X_test)

#Menghitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f"Akurasi Bagging dengan DecisionTree: {accuracy:.4f}")

Akurasi Bagging dengan DecisionTree: 0.8958
